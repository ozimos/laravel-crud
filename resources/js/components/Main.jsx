/* global axios:false */
import React, { Component } from 'react';
import swal from 'sweetalert';
import Product from './Product.jsx';
import ProductEditor from './ProductEditor.jsx';

class Main extends Component {

  constructor() {

    super();
    this.state = {
      products: [],
      displayedProduct: {},
      editorProduct: {},
      canSubmit: false,
      isEdit: false
    };
    this.handleAddProduct = this.handleAddProduct.bind(this);
    this.handleEditProduct = this.handleEditProduct.bind(this);
    this.deleteProduct = this.deleteProduct.bind(this);
    this.handleInput = this.handleInput.bind(this);
    this.handleChecked = this.handleChecked.bind(this);
    this.checkValidity = this.checkValidity.bind(this);
    this.addToEditor = this.addToEditor.bind(this);
  }

  componentDidMount() {
    axios.get('/api/products')
      .then((response) => {
        const products = response.data;
        this.setState({ products, displayedProduct: products[0] });
      });
  }

  handleClick(product) {
    this.setState({ displayedProduct: product });

  }

  handleInput(event) {
    const { editorProduct } = this.state;
    this.setState({
      editorProduct: {
        ...editorProduct,
        [event.target.name]: event.target.value
      },
      canSubmit: this.checkValidity()
    });
  }

  handleChecked(event) {
    const checked = event.target.checked ? 1 : 0;
    const { editorProduct } = this.state;
    this.setState({
      editorProduct: { ...editorProduct, availability: checked }
    });
  }

  checkValidity() {
    const { title, description, price } = this.state.editorProduct;
    return title && description && price;
  }

  addToEditor(product) {
    this.setState({ editorProduct: product, isEdit: true });
  }

  deleteProduct(id) {
    swal({
      title: 'Delete Product',
      text: 'Are you sure you want to delete this product',
      icon: 'warning',
      buttons: [true, 'Delete'],
      dangerMode: true
    })
      .then(() => {
        axios.delete(`api/products/${id}`)
          .then(() => {
            this.setState(prevState => ({
              products: prevState.products
                .filter(product => (product.id !== id)),
            }));
          });
      });
  }

  handleAddProduct(product) {
    product.price = Number(product.price);
    axios.post('api/products/', product, {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
    })
      .then((response) => {
        const savedProduct = response.data;
        this.setState(prevState => ({
          products: [savedProduct, ...prevState.products],
          displayedProduct: savedProduct,
          editorProduct: {}
        }));
      });

  }

  handleEditProduct(product) {
    product.price = Number(product.price);
    axios.put(`api/products/${product.id}`, product, {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
    })
      .then((response) => {
        const savedProduct = response.data;
        this.setState(prevState => ({
          products: prevState.products
            .map(productsItem => (productsItem.id === savedProduct.id
              ? savedProduct : productsItem)),
          displayedProduct: savedProduct,
          editorProduct: {},
          isEdit: false
        }));
      });

  }

  renderProducts() {
    return this.state.products.map(product => (
      <div
        className="flexible"
        key={product.id}
      >
        <button
          onClick={() => this.handleClick(product)}
          type="button"
          className="list-group-item list-group-item-action"
          style={{ width: '80%', borderRight: 'none' }}
        >
          { product.title }
        </button>
        <button
          type="button"
          className="list-group-item list-group-item-action close"
          style={{ width: '10%', color: 'green' }}
          onClick={() => this.addToEditor(product)}
        >
         &#9999;
        </button>
        <button
          type="button"
          className="list-group-item list-group-item-action close"
          style={{ width: '10%', color: 'red' }}
          onClick={() => this.deleteProduct(product.id)}
        >
        &#10006;
        </button>
      </div>
    ));
  }

  render() {
    return (
      <div className="container">
        <div className="row">
          <div className="col-4 product-list">
            <h3> All Products </h3>
            <div className="list-group">
              { this.renderProducts() }
            </div>
          </div>
          <div className="col-8">
            <Product product={this.state.displayedProduct} />
            <hr />
            <ProductEditor
              onAdd={this.handleAddProduct}
              onEdit={this.handleEditProduct}
              product={this.state.editorProduct}
              isEdit={this.state.isEdit}
              handleInput={this.handleInput}
              handleChecked={this.handleChecked}
              canSubmit={Boolean(this.state.canSubmit)}
            />
          </div>
        </div>
      </div>
    );
  }

}

export default Main;
