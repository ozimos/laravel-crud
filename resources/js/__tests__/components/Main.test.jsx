/* global React:false shallow:false toJson:false */

import Main from '../../components/Main';

import { products, displayedProduct, newProduct }
  from '../__mocks__/productsMock';

describe('Main Component', () => {

  const shallowSetup = () => shallow(<Main />);

  it('should render without throwing an error', () => {

    const wrapper = shallowSetup();

    expect(wrapper.find('.product-list')
      .contains(<h3> All Products </h3>)).toBe(true);
  });

  it('renders correctly', () => {
    const wrapper = shallowSetup();
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.setState({ products, displayedProduct });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it(
    'should add a new product '
    + 'when the submit button on the add form is clicked',
    () => {
      const wrapper = shallowSetup();

      const handleAddProductSpy = jest.spyOn(
        wrapper.instance(),
        'handleAddProduct'
      );
      wrapper.instance().handleAddProduct(newProduct);
      expect(handleAddProductSpy).toHaveBeenCalled();
    }
  );

  it('should edit an existing product'
  + ' when the submit button on the edit form is clicked', () => {

    const wrapper = shallowSetup();

    const handleEditProductSpy = jest.spyOn(
      wrapper.instance(),
      'handleEditProduct'
    );
    wrapper.instance().handleEditProduct(displayedProduct);
    expect(handleEditProductSpy).toHaveBeenCalled();
  });

  it('should show delete modal when the delete button is clicked', () => {
    const wrapper = shallowSetup();

    const deleteProductSpy = jest.spyOn(
      wrapper.instance(),
      'deleteProduct'
    );
    wrapper.instance().deleteProduct(products[3].id);
    expect(deleteProductSpy).toHaveBeenCalled();
  });

  it('should delete product when the delete confirm button is clicked', () => {
    const wrapper = shallowSetup();

    const deleteProductSpy = jest.spyOn(
      wrapper.instance(),
      'deleteProduct'
    );
    deleteProductSpy.mockResolvedValue(true);
    wrapper.instance().deleteProduct(products[3].id);
    expect(deleteProductSpy).toHaveBeenCalled();
  });

  it('should show product details when the product button is clicked', () => {
    const wrapper = shallowSetup();

    const handleClickSpy = jest.spyOn(
      wrapper.instance(),
      'handleClick'
    );
    wrapper.instance().handleClick(displayedProduct);
    expect(handleClickSpy).toHaveBeenCalled();
  });

  it('should call `handleInput` when the form input is changed', () => {
    const wrapper = shallowSetup();

    const handleInputSpy = jest.spyOn(
      wrapper.instance(),
      'handleInput'
    );
    wrapper.instance().handleInput({ target: { value: 99 } }, 'abc');
    expect(handleInputSpy).toHaveBeenCalled();
  });

  it('should call `handleChecked` when the form checkbox is changed', () => {
    const wrapper = shallowSetup();

    const handleCheckedSpy = jest.spyOn(
      wrapper.instance(),
      'handleChecked'
    );
    wrapper.instance().handleChecked({ target: { value: 20 } }, 'abc');
    expect(handleCheckedSpy).toHaveBeenCalled();
  });
  it('should check the form validity when a form input is changed', () => {
    const wrapper = shallowSetup();

    const checkValiditySpy = jest.spyOn(
      wrapper.instance(),
      'checkValidity'
    );
    wrapper.instance().checkValidity({ target: { value: 20 } }, 'abc');
    expect(checkValiditySpy).toHaveBeenCalled();
  });
  it('should add a product to the editor when the edit button is clicked',
    () => {
      const wrapper = shallowSetup();

      const addToEditorSpy = jest.spyOn(
        wrapper.instance(),
        'addToEditor'
      );
      wrapper.instance().addToEditor({ target: { value: 20 } }, 'abc');
      expect(addToEditorSpy).toHaveBeenCalled();
    });
});
