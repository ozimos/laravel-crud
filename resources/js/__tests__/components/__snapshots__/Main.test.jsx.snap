// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Component renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-4 product-list"
    >
      <h3>
         All Products 
      </h3>
      <div
        className="list-group"
      />
    </div>
    <div
      className="col-8"
    >
      <Product
        product={Object {}}
      />
      <hr />
      <ProductEditor
        canSubmit={false}
        handleChecked={[Function]}
        handleInput={[Function]}
        isEdit={false}
        onAdd={[Function]}
        onEdit={[Function]}
        product={Object {}}
      />
    </div>
  </div>
</div>
`;

exports[`Main Component renders correctly 2`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-4 product-list"
    >
      <h3>
         All Products 
      </h3>
      <div
        className="list-group"
      >
        <div
          className="flexible"
          key="1"
        >
          <button
            className="list-group-item list-group-item-action"
            onClick={[Function]}
            style={
              Object {
                "borderRight": "none",
                "width": "80%",
              }
            }
            type="button"
          >
            Pressing iron
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "green",
                "width": "10%",
              }
            }
            type="button"
          >
            ✏
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "red",
                "width": "10%",
              }
            }
            type="button"
          >
            ✖
          </button>
        </div>
        <div
          className="flexible"
          key="2"
        >
          <button
            className="list-group-item list-group-item-action"
            onClick={[Function]}
            style={
              Object {
                "borderRight": "none",
                "width": "80%",
              }
            }
            type="button"
          >
            enim
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "green",
                "width": "10%",
              }
            }
            type="button"
          >
            ✏
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "red",
                "width": "10%",
              }
            }
            type="button"
          >
            ✖
          </button>
        </div>
        <div
          className="flexible"
          key="3"
        >
          <button
            className="list-group-item list-group-item-action"
            onClick={[Function]}
            style={
              Object {
                "borderRight": "none",
                "width": "80%",
              }
            }
            type="button"
          >
            molestias
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "green",
                "width": "10%",
              }
            }
            type="button"
          >
            ✏
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "red",
                "width": "10%",
              }
            }
            type="button"
          >
            ✖
          </button>
        </div>
        <div
          className="flexible"
          key="4"
        >
          <button
            className="list-group-item list-group-item-action"
            onClick={[Function]}
            style={
              Object {
                "borderRight": "none",
                "width": "80%",
              }
            }
            type="button"
          >
            est
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "green",
                "width": "10%",
              }
            }
            type="button"
          >
            ✏
          </button>
          <button
            className="list-group-item list-group-item-action close"
            onClick={[Function]}
            style={
              Object {
                "color": "red",
                "width": "10%",
              }
            }
            type="button"
          >
            ✖
          </button>
        </div>
      </div>
    </div>
    <div
      className="col-8"
    >
      <Product
        product={
          Object {
            "availability": 1,
            "description": "Dolor est eligendi sequi sapiente dicta eum.
   Et libero consequuntur non laboriosam officia.
    Inventore velit esse possimus a voluptas mollitia.",
            "id": 4,
            "price": 73,
            "title": "est",
          }
        }
      />
      <hr />
      <ProductEditor
        canSubmit={false}
        handleChecked={[Function]}
        handleInput={[Function]}
        isEdit={false}
        onAdd={[Function]}
        onEdit={[Function]}
        product={Object {}}
      />
    </div>
  </div>
</div>
`;
